server/src/main/java/io/crate/planner/operators/InsertFromValues.java

```diff
@@ -610,23 +611,21 @@ Map<ShardLocation, TReq> resolveAndGroupShardRequests(ShardedRequests<TReq, TIte
                 ShardLocation shardLocation;
                 try {
                     shardLocation = getShardLocation(
-                        index,
+                        partition.asIndexName(),
                         itemAndRoutingAndSourceInfo.item().id(),
                         itemAndRoutingAndSourceInfo.routing(),
-                        clusterService);
+                        operationRouting,
+                        state
```

Comment:
Isn't it save to resolve any state related things while iteration to avoid to operate on stale stats (state has been changed during iteration)?
```suggestion
                        clusterService.operationRouting(),
                        clusterService.state()
```

Reply:
`.state()` is a volatile read and state updates at this point shouldn't influence the result.